src/cryptsetup/cryptsetup.c

```diff
@@ -2416,20 +2393,52 @@ static int run(int argc, char *argv[]) {
                 _cleanup_strv_free_erase_ char **passwords = NULL;
                 for (tries = 0; arg_tries == 0 || tries < arg_tries; tries++) {
+                        _cleanup_(erase_and_freep) void *key_data = NULL;
+                        size_t key_data_size;
+
                         log_debug("Beginning attempt %u to unlock.", tries);
 
                         /* When we were able to acquire multiple keys, let's always process them in this order:
                          *
                          *    1. A key acquired via PKCS#11 or FIDO2 token, or TPM2 chip
-                         *    2. The discovered key: i.e. key_data + key_data_size
-                         *    3. The configured key: i.e. key_file + arg_keyfile_offset + arg_keyfile_size
-                         *    4. The empty password, in case arg_try_empty_password is set
-                         *    5. We enquire the user for a password
+                         *    2. The configured or discovered key, of which both are exclusive and optional
+                         *    3. The empty password, in case arg_try_empty_password is set
+                         *    4. We enquire the user for a password
                          */
 
+                        if (try_discover_key) {
+                                _cleanup_free_ char *bindname = NULL;
+                                const char *fn, *client = "cryptsetup";
+
+                                if (arg_tpm2_device || arg_tpm2_device_auto)
+                                        client = "cryptsetup-tpm2";
+                                else if (arg_fido2_device || arg_fido2_device_auto)
+                                        client = "cryptsetup-fido2";
+                                else if (arg_pkcs11_uri || arg_pkcs11_uri_auto)
+                                        client = "cryptsetup-pkcs11";
+
+                                bindname = make_bindname(volume, client);
+                                if (!bindname)
+                                        return log_oom();
+
+                                /* If a key file is not explicitly specified, search for a key in a well defined
+                                 * search path, and load it. */
+
+                                fn = strjoina(volume, ".key");
+                                r = find_key_file(
+                                                fn,
+                                                STRV_MAKE("/etc/cryptsetup-keys.d", "/run/cryptsetup-keys.d"),
+                                                bindname,
+                                                &key_data, &key_data_size);
+                                if (r < 0)
+                                        return r;
+                                if (r > 0)
+                                        log_debug("Automatically discovered key for volume '%s'.", volume);
+                        }
```

Comment:
this function is large enough, maybe split this entire block out into its own function that first calls determine_mechanism_id(), and then puts together everything and calls find_key_file()

Reply:
Done, moved to `discover_key()`.